File: spring-boot-tipps-reactive/reactive-item-api/src/main/java/io/github/jlmc/reactive/api/v1/exceptionhandler/ControllerExceptionHandler.java
package io.github.jlmc.reactive.api.v1.exceptionhandler;

import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;

@Slf4j
@ControllerAdvice
public class ControllerExceptionHandler  {

    @ExceptionHandler(Exception.class)
    public ResponseEntity<?> handlerUnhatchedException(Exception exception) {
        log.error("Attention: handler Unhatched Exception: {}", exception.getMessage(), exception);

        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                             .body(Problem.of("123-500", exception.getMessage()));

    }

    @ExceptionHandler(RuntimeException.class)
    public ResponseEntity<?> handlerRuntimeException(RuntimeException exception) {
        log.error("Exception caught in handlerRuntimeException : {}", exception.getMessage(), exception );
        return ResponseEntity
                .status(HttpStatus.SERVICE_UNAVAILABLE)
                .body(Problem.of("123-503", exception.getMessage()));
    }
}

File: pnia-kotlin/src/main/kotlin/io/github/jlmc/pniakotlin/api/exceptionhandle/ControllerExceptionHandler.kt
package io.github.jlmc.pniakotlin.api.exceptionhandle

import io.github.jlmc.pniakotlin.domain.services.blacklist.ForbiddenInputValuesException
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.server.ServerWebInputException
import javax.validation.ConstraintViolationException

@ControllerAdvice
class ControllerExceptionHandler {

    private val log = LoggerFactory.getLogger(ControllerExceptionHandler::class.java)

    @ExceptionHandler(Exception::class)
    fun handlerUnhatchedException(exception: Exception): ResponseEntity<*> {
        log.error("Attention: handler Unhatched Exception: {}", exception.message, exception)
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
            .body(Problem("123-500", exception.message))
    }

    @ExceptionHandler(RuntimeException::class)
    fun handlerRuntimeException(exception: RuntimeException): ResponseEntity<*> {
        log.error("Exception caught in handlerRuntimeException : {}", exception.message, exception)
        return ResponseEntity
            .status(HttpStatus.INTERNAL_SERVER_ERROR)
            .body(Problem("500", exception.message))
    }

    @ExceptionHandler(
        ConstraintViolationException::class,
        ServerWebInputException::class
    )
    fun handlerValidationExceptions(exception: Exception): ResponseEntity<*> {
        log.error("Exception caught in handlerValidationExceptions : {}", exception.message, exception)
        return ResponseEntity
            .status(HttpStatus.BAD_REQUEST)
            .body(Problem("400", exception.message))
    }

    @ExceptionHandler(ForbiddenInputValuesException::class)
    fun handlerForbiddenInputValuesException(exception: ForbiddenInputValuesException): ResponseEntity<*> {
        log.error("Exception caught in handlerForbiddenInputValuesException : {}", exception.message, exception)
        return ResponseEntity
            .status(HttpStatus.BAD_REQUEST)
            .body(
                Problem(
                    "400",
                    "Unaccepted input values " + exception.invalidValues.joinToString(", ", "[", "]")
                )
            )
    }
}
